src/isaac_system.cpp

```diff
@@ -0,0 +1,178 @@
+//      notice, this list of conditions and the following disclaimer in the
+//      documentation and/or other materials provided with the distribution.
+//
+//    * Neither the name of the PickNik Inc. nor the names of its
+//      contributors may be used to endorse or promote products derived from
+//      this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/* Author: Jafar Abdi */
+#include <algorithm>
+#include <cmath>
+#include <iterator>
+#include <limits>
+#include <set>
+#include <string>
+#include <vector>
+
+#include <isaac_ros2_control/isaac_system.hpp>
+#include <rclcpp/executors.hpp>
+
+namespace isaac_ros2_control
+{
+
+CallbackReturn IsaacSystem::on_init(const hardware_interface::HardwareInfo& info)
+{
+  if (hardware_interface::SystemInterface::on_init(info) != CallbackReturn::SUCCESS)
+  {
+    return CallbackReturn::ERROR;
+  }
+
+  // Initialize storage for all joints' standard interfaces, regardless of their existence and set all values to nan
+  joint_commands_.resize(standard_interfaces_.size());
+  joint_states_.resize(standard_interfaces_.size());
+  for (auto i = 0u; i < standard_interfaces_.size(); i++)
+  {
+    joint_commands_[i].resize(info_.joints.size(), std::numeric_limits<double>::quiet_NaN());
+    joint_states_[i].resize(info_.joints.size(), 0.0);
+  }
+
+  const auto get_hardware_parameter = [this](const std::string& parameter_name, const std::string& default_value) {
+    if (auto it = info_.hardware_parameters.find(parameter_name); it != info_.hardware_parameters.end())
+    {
+      return it->second;
+    }
+    return default_value;
+  };
+
+  node_ = rclcpp::Node::make_shared("isaac_ros2_control");
+  isaac_joint_commands_publisher_ = node_->create_publisher<sensor_msgs::msg::JointState>(
+      get_hardware_parameter("joint_commands_topic", "/joint_command"), rclcpp::QoS(1));
+  isaac_joint_states_subscriber_ = node_->create_subscription<sensor_msgs::msg::JointState>(
+      get_hardware_parameter("joint_states_topic", "/isaac_joint_states"), rclcpp::SensorDataQoS(),
+      [this](const sensor_msgs::msg::JointState::SharedPtr joint_state) { latest_joint_state_ = *joint_state; });
```

Comment:
This is fine for now but we need to document how to setup Isaac/this hardware interface to communicate together via ROS topics.

Reply:
It would also be best if we measured the publish rate and warn users if the controller is running too fast/slow.